### PR TITLE
CT descriptor: remove generic from blinding key

### PR DESCRIPTION
**Pull request by LeoComandini**
The generic was only used by the `Bare` variant, which was inconsistent with its corresponding "secret" variant, `View`.

The generic could be used to handle the conversion from `ConfidentialDescriptor<DescriptorPublicKey>` to
`Descriptor<DefiniteDescriptorKey>`, specifically making sure that the descriptor blinding key does not have wildcards.

For the `Bare` variant this could be done, but this cannot happen for the `View` variant, since we don't have the definite version of `DescriptorSecretKey`.

The proposed solution consists in:
* Remove generic from descriptor blinding key (this commit)
* Add `ConfidentialDescriptor::at_derivation_index` that "removes" wildcards from both the "bitcoin" descriptor and the descriptor blinding key. This makes deriving an address from a CT descriptor with wildcards less cumbersome.
* Change `Key::to_public_key(&self, secp, spk)` to return a `Result` and error if `Key` has some wildcards. This will fix the case `ct(xprv/*,desc)` which is currently broken.
* Add test cases for blinding keys with wildcards
* Add test vectors with blinding keys with wildcards to ELIP-150

### CODE REVIEW

**Comment by apoelstra** [comment]
Yeah, this solution seems good to me. It's not great but nothing I could come up with was great either.

I'd like to fix CI though before moving forward on this.

**Comment by LeoComandini** [comment]
> fix CI

Fixed by @RCasatta in #60 , I'll rebase on top of that once merged

**Comment by LeoComandini** [comment]
Proposed additional test vectors for ELIP-150 
(will post there once this gets a round of review)

View Descriptor with wildcard: ct(xprv9s21ZrQH143K28NgQ7bHCF61hy9VzwquBZvpzTwXLsbmQLRJ6iV9k2hUBRt5qzmBaSpeMj5LdcsHaXJvM7iFEivPryRcL8irN7Na9p65UUb/*,elwpkh(xpub661MyMwAqRbcEcT9W98HZP2kFzyzQQZkYnrRnrM8uD8kH8kSeFoQHq1x2iihLgC6PXGy5LrjCL66uSNhJ8pwjfx2rMUTLWuRMns2EG9xnjs/*))#
    
    Index: 1,
    Descriptor blinding private key: xprv9s21ZrQH143K28NgQ7bHCF61hy9VzwquBZvpzTwXLsbmQLRJ6iV9k2hUBRt5qzmBaSpeMj5LdcsHaXJvM7iFEivPryRcL8irN7Na9p65UUb/*
    Confidential address: el1qqf6690fpw2y00hv5a84zsydjgztg2089d5xnll4k4cstzn63uvgudd907qpvlvvwd5ym9gx7j0v46elf23kfxunucm6ejjyk0
    Unconfidential address: ert1qkjhlqqk0kx8x6zdj5r0f8k2avl54gmyn7qjk2k

Non-View Descriptor with wildcard: ct(xpub661MyMwAqRbcEcT9W98HZP2kFzyzQQZkYnrRnrM8uD8kH8kSeFoQHq1x2iihLgC6PXGy5LrjCL66uSNhJ8pwjfx2rMUTLWuRMns2EG9xnjs/*,elwpkh(xpub661MyMwAqRbcEcT9W98HZP2kFzyzQQZkYnrRnrM8uD8kH8kSeFoQHq1x2iihLgC6PXGy5LrjCL66uSNhJ8pwjfx2rMUTLWuRMns2EG9xnjs/*))#
    
    Index: 1,
    Descriptor blinding public key: xpub661MyMwAqRbcEcT9W98HZP2kFzyzQQZkYnrRnrM8uD8kH8kSeFoQHq1x2iihLgC6PXGy5LrjCL66uSNhJ8pwjfx2rMUTLWuRMns2EG9xnjs/*
    Confidential address: el1qqf6690fpw2y00hv5a84zsydjgztg2089d5xnll4k4cstzn63uvgudd907qpvlvvwd5ym9gx7j0v46elf23kfxunucm6ejjyk0
    Unconfidential address: ert1qkjhlqqk0kx8x6zdj5r0f8k2avl54gmyn7qjk2k

File: src/confidential/mod.rs [hunk 1]
```diff
@@ -82,6 +86,51 @@ impl<Pk: MiniscriptKey, T: Extension> Descriptor<Pk, T> {
...
+            Key::View(k) => Key::View(match k {
+                // Consider implementing DescriptorSecretKey::at_derivation_index
+                DescriptorSecretKey::Single(_) => k,
+                DescriptorSecretKey::XPrv(xprv) => {
+                    let derivation_path = match xprv.wildcard {
+                        Wildcard::None => xprv.derivation_path,
+                        Wildcard::Unhardened => xprv.derivation_path.into_child(
+                            bip32::ChildNumber::from_normal_idx(index)
+                                .ok()
+                                .ok_or(ConversionError::HardenedChild)?,
```
Comment (RCasatta) [inline src/confidential/mod.rs:132]: `map_err` ?

**Review by RCasatta — COMMENTED** [review]
LGTM

**Comment by LeoComandini** [comment]
> > fix CI
> 
> Fixed by @RCasatta in #60 , I'll rebase on top of that once merged

@apoelstra rebased and CI is now green

**Review by apoelstra — APPROVED** [review]
ACK f4be6afefc7a46af96fda083b1be7bc3fb261498

**Comment by apoelstra** [comment]
Looks good to me. Obviously there are still some open API questions about doing derivations but the result of this is much better than the existing code.

**Pull request merged**